obstore/python/obstore/fsspec.py

```diff
@@ -0,0 +1,181 @@
+        starts: List[int],
+        ends: List[int],
+        max_gap=None,
+        batch_size=None,
+        on_error="return",
+        **kwargs,
+    ):
+        if not len(paths) == len(starts) == len(ends):
+            raise ValueError
+
+        per_file_requests: Dict[str, List[Tuple[int, int, int]]] = defaultdict(list)
+        for idx, (path, start, end) in enumerate(zip(paths, starts, ends)):
+            per_file_requests[path].append((start, end, idx))
+
+        futs: List[Coroutine[Any, Any, List[bytes]]] = []
+        for path, ranges in per_file_requests.items():
+            offsets = [r[0] for r in ranges]
+            ends = [r[1] for r in ranges]
+            fut = obs.get_ranges_async(self.store, path, starts=offsets, ends=ends)
+            futs.append(fut)
+
+        result = await asyncio.gather(*futs)
+
+        output_buffers: List[bytes] = [b""] * len(paths)
+        for per_file_request, buffers in zip(per_file_requests.items(), result):
+            path, ranges = per_file_request
+            for buffer, ranges_ in zip(buffers, ranges):
+                initial_index = ranges_[2]
+                output_buffers[initial_index] = buffer.as_bytes()
+
+        return output_buffers
+
+    async def _put_file(self, lpath, rpath, **kwargs):
+        with open(lpath, "rb") as f:
+            await obs.put_async(self.store, rpath, f)
+
+    async def _get_file(self, rpath, lpath, **kwargs):
+        with open(lpath, "wb") as f:
+            resp = await obs.get_async(self.store, rpath)
+            async for buffer in resp.stream():
+                f.write(buffer)
+
+    async def _info(self, path, **kwargs):
+        head = await obs.head_async(self.store, path)
+        return {
+            # Required of `info`: (?)
+            "name": head["path"],
+            "size": head["size"],
+            "type": "directory" if head["path"].endswith("/") else "file",
+            # Implementation-specific keys
+            "e_tag": head["e_tag"],
+            "last_modified": head["last_modified"],
+            "version": head["version"],
+        }
+
+    async def _ls(self, path, detail=True, **kwargs):
+        result = await obs.list_with_delimiter_async(self.store, path)
+        objects = result["objects"]
+        prefs = result["common_prefixes"]
+        if detail:
+            return [
+                {
+                    "name": object["path"],
+                    "size": object["size"],
+                    "type": "file",
+                    "ETag": object["e_tag"],
+                }
+                for object in objects
+            ] + [{"name": object, "size": 0, "type": "directory"} for object in prefs]
+        else:
+            return sorted([object["path"] for object in objects] + prefs)
+
+    def _open(self, path, mode="rb", **kwargs):
+        """Return raw bytes-mode file-like from the file-system"""
+        out = BufferedFileSimple(self, path, mode)
+        return out
```

Comment:
Is this what's going to have a default implementation when https://github.com/fsspec/filesystem_spec/pull/1732 is merged? Or this is separate?

Reply:
It could have been done with that PR, once release. 

I was worried about passing Buffers around rather than bytes and/or whether read buffering might be implemented on the rust side (without copies). I can confirm that AbstractBufferedFile works directly including the fsspec change. 

However, we *will* end up making a class to deal with writing data unless we only every write whole memory buffers in a single go as opposed to multi-part.

Reply:
> However, we _will_ end up making a class to deal with writing data unless we only every write whole memory buffers in a single go as opposed to multi-part.

We can wrap a `BufWriter` similarly to what we do now with `BufReader`, which will use a multi part upload under the hood.